crates/bevy_render/src/view/window/screenshot.rs

```diff
@@ -0,0 +1,314 @@
+                polygon_mode: wgpu::PolygonMode::Fill,
+                conservative: false,
+                unclipped_depth: false,
+            },
+            depth_stencil: None,
+            multisample: wgpu::MultisampleState::default(),
+            fragment: Some(FragmentState {
+                shader: SCREENSHOT_SHADER_HANDLE.typed(),
+                entry_point: Cow::Borrowed("fs_main"),
+                shader_defs: vec![],
+                targets: vec![Some(wgpu::ColorTargetState {
+                    format: key,
+                    blend: None,
+                    write_mask: wgpu::ColorWrites::ALL,
+                })],
+            }),
+        }
+    }
+}
+
+pub struct ScreenshotPreparedState {
+    pub texture: Texture,
+    pub buffer: Buffer,
+    pub bind_group: BindGroup,
+    pub pipeline_id: CachedRenderPipelineId,
+}
+
+pub(crate) fn submit_screenshot_commands(world: &World, encoder: &mut CommandEncoder) {
+    let windows = world.resource::<ExtractedWindows>();
+    let pipelines = world.resource::<PipelineCache>();
+
+    for window in windows.values() {
+        if let Some(memory) = &window.screenshot_memory {
+            let width = window.physical_width;
+            let height = window.physical_height;
+            let texture_format = window.swap_chain_texture_format.unwrap();
+
+            encoder.copy_texture_to_buffer(
+                memory.texture.as_image_copy(),
+                wgpu::ImageCopyBuffer {
+                    buffer: &memory.buffer,
+                    layout: crate::view::screenshot::layout_data(width, height, texture_format),
+                },
+                Extent3d {
+                    width,
+                    height,
+                    ..Default::default()
+                },
+            );
+            if let Some(pipeline) = pipelines.get_render_pipeline(memory.pipeline_id) {
+                let true_swapchain_texture_view = window
+                    .swap_chain_texture
+                    .as_ref()
+                    .unwrap()
+                    .texture
+                    .create_view(&Default::default());
+                let mut pass = encoder.begin_render_pass(&wgpu::RenderPassDescriptor {
+                    label: Some("screenshot_to_screen_pass"),
+                    color_attachments: &[Some(wgpu::RenderPassColorAttachment {
+                        view: &true_swapchain_texture_view,
+                        resolve_target: None,
+                        ops: wgpu::Operations {
+                            load: wgpu::LoadOp::Load,
+                            store: true,
+                        },
+                    })],
+                    depth_stencil_attachment: None,
+                });
+                pass.set_pipeline(pipeline);
+                pass.set_bind_group(0, &memory.bind_group, &[]);
+                pass.draw(0..3, 0..1);
+            }
+        }
+    }
+}
+
+pub(crate) fn collect_screenshots(world: &mut World) {
```

Comment:
Can you explain a little bit what this is doing? I'm not entirely sure I understand what's going on here.

Reply:
This function basically loops over all the windows, sees which ones have had a screenshot taken, and creates async tasks per window with a screenshot to copy that memory from the GPU to the CPU, and eventually calls the relevant screenshot callback.

In a way it's "collecting" the screenshots that were taken to the CPU to continue processing